include/unifex/bulk_with_stop_source.hpp

```diff
@@ -0,0 +1,231 @@
+
+    template<typename Source2>
+    explicit type(Source2&& source) : source_((Source2&&)source)
+    {}
+
+    template(typename Self, typename Receiver)
+        (requires
+            same_as<remove_cvref_t<Self>, type> AND
+            sender_to<member_t<Self, Source>, stop_source_receiver<operation<Source, Receiver>, remove_cvref_t<Receiver>>>)
+    friend auto tag_invoke(tag_t<unifex::connect>, Self&& self, Receiver&& r)
+        noexcept(
+            std::is_nothrow_constructible_v<Source, member_t<Self, Source>> &&
+            std::is_nothrow_constructible_v<remove_cvref_t<Receiver>, Receiver>) {
+
+        return operation<Source, Receiver>(
+            static_cast<Self&&>(self).source_, std::forward<Receiver>(r));
+    }
+
+private:
+    UNIFEX_NO_UNIQUE_ADDRESS Source source_;
+};
+
+template<typename Pred, typename Receiver, typename Token>
+struct _stop_source_operation_callback;
+template<typename Pred, typename Receiver, typename Token>
+struct _stop_source_operation_callback {
+    _stop_source_operation_callback(operation<Pred, Receiver>& op, Token&& token) :
+        op_{op}, token_{token} {}
+    void operator()() {
+        op_.stop_source_.request_stop();
+    }
+    operation<Pred, Receiver>& op_;
+    Token token_;
+};
+template<typename Pred, typename Receiver>
+struct _stop_source_operation_callback<Pred, Receiver, unstoppable_token> {
+    _stop_source_operation_callback(operation<Pred, Receiver>& /*op*/, unstoppable_token&& /*token*/)  {}
+    void operator()() {}
+};
+
+template<typename Pred, typename Receiver>
+struct _stop_source_operation<Pred, Receiver>::type {
+    type(Pred&& pred, Receiver&& r) :
+        stop_callback_(*this, unifex::get_stop_token(r)),
+        predOp_(
+            unifex::connect(
+                (Pred&&)(pred),
+                stop_source_receiver<operation<Pred, Receiver>, remove_cvref_t<Receiver>>{
+                    *this,
+                    static_cast<Receiver&&>(r)})),
+        stop_source_{} {
+    }
+
+    void start() noexcept {
+        unifex::start(predOp_);
+    }
+
+    UNIFEX_NO_UNIQUE_ADDRESS _stop_source_operation_callback<
+            Pred, Receiver, remove_cvref_t<decltype(unifex::get_stop_token(std::declval<Receiver>()))>>
+        stop_callback_;
+    connect_result_t<
+        Pred,
+        stop_source_receiver<operation<Pred, Receiver>, remove_cvref_t<Receiver>>>
+        predOp_;
+    UNIFEX_NO_UNIQUE_ADDRESS unifex::inplace_stop_source stop_source_;
+};
+
+
+struct _fn {
+
+    template(typename Source)
+        (requires typed_bulk_sender<Source>)
+    auto operator()(Source&& s) const
+        noexcept(std::is_nothrow_constructible_v<remove_cvref_t<Source>, Source>)
+        -> stop_source_sender<remove_cvref_t<Source>> {
+        return stop_source_sender<remove_cvref_t<Source>>{(Source&&)s};
+    }
+};
+
+} // namespace _bulk_with_stop_source
+
+inline constexpr _bulk_with_stop_source::_fn bulk_with_stop_source{};
```

Comment:
I have a feeling this algorithm could be made more general to be applicable to all senders, not just bulk senders.

e.g. Say a `let_stop_source(fn)` that returns a sender that constructs an inplace_stop_source and invokes `fn` with an lvalue reference to it, which is then responsible for returning a sender which produces the result of the expression. It then connects the returned sender with a receiver that has the inplace_stop_token corresponding to that inplace_stop_source and also hooks up the incoming receiver's stop token to also request stop on the inplace_stop_source.

This could then be combined with a bulk operation as follows.
```
let_stop_source([&](inplace_stop_source& stopSource) {
  return bulk_join(
    bulk_transform(
      bulk_schedule(n),
      [&](size_t i) {
        if (shouldStop(i)) {
          stopSource.request_stop();
        }
      }));
});
```

Reply:
We could even make this a little more generic by having, say, a `let_value_from(factory, fn)` that is like `let_value()` but that constructs the value using `factor()` instead of from another sender.
This would allow constructing in-place an immovable object like `inplace_stop_source`.

It's possible that my experiments on sender/receiver async cleanup with async rvo might directly yield the ability to use `let_value()` with a `generate(fn)` algorithm to achieve the same. But that's something for another day...

Reply:
I think you're right. I don't tend to think in terms of let but it does seem likely to be cleaner in this case. I should do that more often - putting values in the data stream isn't always the right approach once you have good structure.

Reply:
I left it as the stop_source injecting version of let. We can make it more flexible later if necessary.